File: java/0015-3sum.java
import java.util.*;

/**
 * LeetCode Problem 0015: 3Sum
 *
 * Problem:
 * Find all unique triplets in the array which gives the sum of zero.
 *
 * Note: The solution set must not contain duplicate triplets.
 *
 * Examples:
 * 1. Input: nums = [-1,0,1,2,-1,-4], Output: [[-1,-1,2],[-1,0,1]]
 * 2. Input: nums = [0], Output: []
 */
public class ThreeSum {

    /**
     * Q1: Sorting + Two Pointer (most efficient)
     *
     * Time Complexity: O(n^2)
     * Space Complexity: O(1) or O(n) depending on sorting
     */
    public static List<List<Integer>> threeSumSorting(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> result = new ArrayList<>();
        int n = nums.length;

        for (int i = 0; i < n - 2; i++) {
            if (i > 0 && nums[i] == nums[i - 1]) continue;
            if (nums[i] + nums[i + 1] + nums[i + 2] > 0) break;
            if (nums[i] + nums[n - 2] + nums[n - 1] < 0) continue;

            int target = -nums[i];
            int left = i + 1, right = n - 1;

            while (left < right) {
                int sum = nums[left] + nums[right];
                if (sum == target) {
                    result.add(Arrays.asList(nums[i], nums[left], nums[right]));
                    while (left < right && nums[left] == nums[left + 1]) left++;
                    while (left < right && nums[right] == nums[right - 1]) right--;
                    left++;
                    right--;
                } else if (sum < target) {
                    left++;
                } else {
                    right--;
                }
            }
        }

        return result;
    }

    /**
     * Q2: HashSet based approach
     *
     * Time Complexity: O(n^2)
     * Space Complexity: O(n)
     */
    public static List<List<Integer>> threeSumHashset(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> result = new ArrayList<>();
        Set<List<Integer>> resultSet = new HashSet<>();

        for (int i = 0; i < nums.length - 2; i++) {
            Set<Integer> seen = new HashSet<>();
            int target = -nums[i];

            for (int j = i + 1; j < nums.length; j++) {
                int complement = target - nums[j];
                if (seen.contains(complement)) {
                    List<Integer> triplet = Arrays.asList(nums[i], complement, nums[j]);
                    resultSet.add(triplet);
                }
                seen.add(nums[j]);
            }
        }

        result.addAll(resultSet);
        return result;
    }

    /**
     * Q3: Two pointer with different implementation
     *
     * Time Complexity: O(n^2)
     * Space Complexity: O(1)
     */
    public static List<List<Integer>> threeSumTwoPointer(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> result = new ArrayList<>();

        for (int i = 0; i < nums.length - 2; i++) {
            if (nums[i] > 0) break;
            if (i > 0 && nums[i] == nums[i - 1]) continue;

            int left = i + 1, right = nums.length - 1;
            int target = -nums[i];

            while (left < right) {
                int sum = nums[left] + nums[right];
                if (sum == target) {
                    result.add(Arrays.asList(nums[i], nums[left], nums[right]));

                    while (left < right && nums[left + 1] == nums[left]) left++;
                    while (left < right && nums[right - 1] == nums[right]) right--;

                    left++;
                    right--;
                } else if (sum < target) {
                    left++;
                } else {
                    right--;
                }
            }
        }

        return result;
    }

    /**
     * Q4: Using Set for result deduplication
     *
     * Time Complexity: O(n^2)
     * Space Complexity: O(n)
     */
    public static List<List<Integer>> threeSumSetDedup(int[] nums) {
        Arrays.sort(nums);
        Set<List<Integer>> resultSet = new HashSet<>();

        for (int i = 0; i < nums.length - 2; i++) {
            int left = i + 1, right = nums.length - 1;

            while (left < right) {
                int sum = nums[i] + nums[left] + nums[right];
                if (sum == 0) {
                    resultSet.add(Arrays.asList(nums[i], nums[left], nums[right]));
                    left++;
                    right--;
                } else if (sum < 0) {
                    left++;
                } else {
                    right--;
                }
            }
        }

        return new ArrayList<>(resultSet);
    }

    /**
     * Q5: Index-based approach
     *
     * Time Complexity: O(n^2)
     * Space Complexity: O(1)
     */
    public static List<List<Integer>> threeSumIndex(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> result = new ArrayList<>();
        int n = nums.length;

        for (int i = 0; i < n - 2; i++) {
            if (i > 0 && nums[i] == nums[i - 1]) continue;

            int left = i + 1, right = n - 1;

            while (left < right) {
                int sum = nums[i] + nums[left] + nums[right];

                if (sum == 0) {
                    result.add(Arrays.asList(nums[i], nums[left], nums[right]));

                    while (left + 1 < right && nums[left] == nums[left + 1]) left++;
                    while (right - 1 > left && nums[right] == nums[right - 1]) right--;

                    left++;
                    right--;
                } else if (sum < 0) {
                    left++;
                } else {
                    right--;
                }
            }
        }

        return result;
    }

    /**
     * Q6: Optimized with early termination
     *
     * Time Complexity: O(n^2)
     * Space Complexity: O(1)
     */
    public static List<List<Integer>> threeSumOptimized(int[] nums) {
        if (nums == null || nums.length < 3) return new ArrayList<>();

        Arrays.sort(nums);
        List<List<Integer>> result = new ArrayList<>();

        for (int i = 0; i < nums.length - 2; i++) {
            if (nums[i] > 0) break;
            if (i > 0 && nums[i] == nums[i - 1]) continue;

            twoSum(nums, i, result);
        }

        return result;
    }

    private static void twoSum(int[] nums, int i, List<List<Integer>> result) {
        int left = i + 1, right = nums.length - 1;
        int target = -nums[i];

        while (left < right) {
            int sum = nums[left] + nums[right];
            if (sum == target) {
                result.add(Arrays.asList(nums[i], nums[left], nums[right]));
                while (left < right && nums[left] == nums[left + 1]) left++;
                while (left < right && nums[right] == nums[right - 1]) right--;
                left++;
                right--;
            } else if (sum < target) {
                left++;
            } else {
                right--;
            }
        }
    }

    /**
     * Q7: Using TreeSet for sorted unique results
     *
     * Time Complexity: O(n^2 log n)
     * Space Complexity: O(n)
     */
    public static List<List<Integer>> threeSumTreeSet(int[] nums) {
        Arrays.sort(nums);
        TreeSet<List<Integer>> resultSet = new TreeSet<>((a, b) -> {
            for (int i = 0; i < 3; i++) {
                if (!a.get(i).equals(b.get(i))) {
                    return a.get(i) - b.get(i);
                }
            }
            return 0;
        });

        for (int i = 0; i < nums.length - 2; i++) {
            int left = i + 1, right = nums.length - 1;

            while (left < right) {
                int sum = nums[i] + nums[left] + nums[right];
                if (sum == 0) {
                    resultSet.add(Arrays.asList(nums[i], nums[left], nums[right]));
                    left++;
                    right--;
                } else if (sum < 0) {
                    left++;
                } else {
                    right--;
                }
            }
        }

        return new ArrayList<>(resultSet);
    }

    /**
     * Q8: Optimized with bounds checking
     *
     * Time Complexity: O(n^2)
     * Space Complexity: O(1)
     */
    public static List<List<Integer>> threeSumBounds(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> result = new ArrayList<>();

        for (int i = 0; i < nums.length - 2; i++) {
            if (nums[i] > 0) break;
            if (i > 0 && nums[i] == nums[i - 1]) continue;

            if (nums[i] + nums[i + 1] + nums[i + 2] > 0) break;
            if (nums[i] + nums[nums.length - 2] + nums[nums.length - 1] < 0)
                continue;

            int target = -nums[i];
            int left = i + 1, right = nums.length - 1;

            while (left < right) {
                int sum = nums[left] + nums[right];
                if (sum == target) {
                    result.add(Arrays.asList(nums[i], nums[left], nums[right]));
                    while (left < right && nums[left] == nums[left + 1]) left++;
                    while (left < right && nums[right] == nums[right - 1]) right--;
                    left++;
                    right--;
                } else if (sum < target) {
                    left++;
                } else {
                    right--;
                }
            }
        }

        return result;
    }

    /**
     * Q9: Stream-based approach
     *
     * Time Complexity: O(n^2)
     * Space Complexity: O(n)
     */
    public static List<List<Integer>> threeSumStream(int[] nums) {
        Arrays.sort(nums);
        Set<List<Integer>> resultSet = new HashSet<>();

        for (int i = 0; i < nums.length - 2; i++) {
            Set<Integer> seen = new HashSet<>();
            int target = -nums[i];

            for (int j = i + 1; j < nums.length; j++) {
                int complement = target - nums[j];
                if (seen.contains(complement)) {
                    resultSet.add(Arrays.asList(nums[i], complement, nums[j]));
                }
                seen.add(nums[j]);
            }
        }

        return new ArrayList<>(resultSet);
    }

    /**
     * Q10: Clean implementation with helper
     *
     * Time Complexity: O(n^2)
     * Space Complexity: O(1)
     */
    public static List<List<Integer>> threeSumClean(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> result = new ArrayList<>();

        for (int i = 0; i < nums.length - 2; i++) {
            if (i > 0 && nums[i] == nums[i - 1]) continue;
            if (nums[i] > 0) break;

            int target = -nums[i];
            int left = i + 1, right = nums.length - 1;

            while (left < right) {
                int sum = nums[left] + nums[right];
                if (sum == target) {
                    result.add(Arrays.asList(nums[i], nums[left], nums[right]));
                    skipDuplicates(nums, left, right, result);
                    left++;
                    right--;
                } else if (sum < target) {
                    left++;
                } else {
                    right--;
                }
            }
        }

        return result;
    }

    private static void skipDuplicates(int[] nums, int left, int right,
                                       List<List<Integer>> result) {
        while (left < right && nums[left] == nums[left + 1]) left++;
        while (left < right && nums[right] == nums[right - 1]) right--;
    }

    // Test cases and main function
    public static void main(String[] args) {
        int[][] testCases = {
                {-1, 0, 1, 2, -1, -4},
                {0},
                {-2, 0, 1, 1, 2},
                {0, 0, 0, 0},
        };

        List<List<Integer>>[] expectedResults = new List[]{
                Arrays.asList(
                        Arrays.asList(-1, -1, 2),
                        Arrays.asList(-1, 0, 1)
                ),
                new ArrayList<>(),
                Arrays.asList(
                        Arrays.asList(-2, 0, 2),
                        Arrays.asList(-2, 1, 1)
                ),
                Arrays.asList(Arrays.asList(0, 0, 0))
        };

        java.lang.reflect.Method[] methods = ThreeSum.class.getDeclaredMethods();
        List<java.lang.reflect.Method> testMethods = new ArrayList<>();

        for (java.lang.reflect.Method method : methods) {
            if (method.getName().startsWith("threeSum") &&
                method.getReturnType() == List.class) {
                testMethods.add(method);
            }
        }

        Collections.sort(testMethods, Comparator.comparing(java.lang.reflect.Method::getName));

        int questionNum = 1;
        for (java.lang.reflect.Method method : testMethods) {
            System.out.println("\nQ" + questionNum + ": " + method.getName());
            System.out.println("-".repeat(70));

            boolean allPassed = true;
            for (int i = 0; i < testCases.length; i++) {
                try {
                    List<List<Integer>> result = (List<List<Integer>>)
                        method.invoke(null, (Object) testCases[i]);

                    result.sort((a, b) -> {
                        for (int j = 0; j < 3; j++) {
                            if (!a.get(j).equals(b.get(j))) {
                                return a.get(j) - b.get(j);
                            }
                        }
                        return 0;
                    });

                    List<List<Integer>> expected = expectedResults[i];
                    expected.sort((a, b) -> {
                        for (int j = 0; j < 3; j++) {
                            if (!a.get(j).equals(b.get(j))) {
                                return a.get(j) - b.get(j);
                            }
                        }
                        return 0;
                    });

                    boolean passed = result.equals(expected);
                    allPassed = allPassed && passed;
                    String status = passed ? "✓" : "✗";
                    System.out.println("  " + status + " threeSum(" +
                        Arrays.toString(testCases[i]) + ") = " + result);
                } catch (Exception e) {
                    System.out.println("  ✗ Error: " + e.getMessage());
                    allPassed = false;
                }
            }

            if (allPassed) {
                System.out.println("  All tests passed!");
            }

            questionNum++;
        }
    }
}

File: python/0015-3sum.py
"""
LeetCode Problem 0015: 3Sum

Problem:
Find all unique triplets in the array which gives the sum of zero.

Note: The solution set must not contain duplicate triplets, and you may return
the solution set in any order.

Examples:
1. Input: nums = [-1,0,1,2,-1,-4], Output: [[-1,-1,2],[-1,0,1]]
2. Input: nums = [0], Output: []
3. Input: nums = [-2,0,1,1,2], Output: [[-2,0,2],[-2,1,1]]

Constraints:
- 0 <= nums.length <= 3000
- -10^5 <= nums[i] <= 10^5
"""

# Q1: Sorting + Two Pointer (most efficient)
def threeSum_sorting(nums: list[int]) -> list[list[int]]:
    """
    Time Complexity: O(n^2)
    Space Complexity: O(1) or O(n) depending on sorting algorithm

    Approach: Sort array, then for each element, use two pointers to find pairs
    that sum to the negative of current element.
    """
    nums.sort()
    result = []
    n = len(nums)

    for i in range(n - 2):
        # Skip duplicate values
        if i > 0 and nums[i] == nums[i - 1]:
            continue

        # If smallest possible sum is positive, break
        if nums[i] + nums[i + 1] + nums[i + 2] > 0:
            break

        # If largest possible sum is negative, continue
        if nums[i] + nums[n - 2] + nums[n - 1] < 0:
            continue

        target = -nums[i]
        left, right = i + 1, n - 1

        while left < right:
            current_sum = nums[left] + nums[right]
            if current_sum == target:
                result.append([nums[i], nums[left], nums[right]])
                # Skip duplicates
                while left < right and nums[left] == nums[left + 1]:
                    left += 1
                while left < right and nums[right] == nums[right - 1]:
                    right -= 1
                left += 1
                right -= 1
            elif current_sum < target:
                left += 1
            else:
                right -= 1

    return result


# Q2: Hash set based approach
def threeSum_hashset(nums: list[int]) -> list[list[int]]:
    """
    Time Complexity: O(n^2)
    Space Complexity: O(n)

    Approach: For each pair, use hash set to check if complement exists.
    """
    nums.sort()
    result = []
    n = len(nums)

    for i in range(n - 2):
        if i > 0 and nums[i] == nums[i - 1]:
            continue

        if nums[i] > 0:
            break

        seen = set()
        for j in range(i + 1, n):
            complement = -nums[i] - nums[j]
            if complement in seen:
                result.append([nums[i], complement, nums[j]])
                while j + 1 < n and nums[j] == nums[j + 1]:
                    j += 1
            seen.add(nums[j])

    return result


# Q3: Two pointer with no sorting optimization
def threeSum_two_pointer(nums: list[int]) -> list[list[int]]:
    """
    Time Complexity: O(n^2)
    Space Complexity: O(1)

    Approach: Similar to Q1 but with different implementation style.
    """
    nums.sort()
    result = []

    for i in range(len(nums) - 2):
        if nums[i] > 0:
            break

        if i > 0 and nums[i] == nums[i - 1]:
            continue

        left, right = i + 1, len(nums) - 1
        target = -nums[i]

        while left < right:
            total = nums[left] + nums[right]
            if total == target:
                result.append([nums[i], nums[left], nums[right]])

                while left < right and nums[left + 1] == nums[left]:
                    left += 1
                while left < right and nums[right - 1] == nums[right]:
                    right -= 1

                left += 1
                right -= 1
            elif total < target:
                left += 1
            else:
                right -= 1

    return result


# Q4: Dictionary-based approach
def threeSum_dict(nums: list[int]) -> list[list[int]]:
    """
    Time Complexity: O(n^2)
    Space Complexity: O(n)

    Approach: Build a dictionary and use it for lookups.
    """
    nums.sort()
    result = []
    seen = set()

    for i in range(len(nums) - 2):
        if nums[i] in seen:
            continue
        seen.add(nums[i])

        pairs = {}
        for j in range(i + 1, len(nums)):
            complement = -nums[i] - nums[j]
            if complement in pairs:
                triplet = tuple(sorted([nums[i], complement, nums[j]]))
                result.append(list(triplet))
            pairs[nums[j]] = j

    return result


# Q5: Using set for triplet deduplication
def threeSum_set_dedup(nums: list[int]) -> list[list[int]]:
    """
    Time Complexity: O(n^2)
    Space Complexity: O(n)

    Approach: Use set to automatically handle duplicates in results.
    """
    nums.sort()
    result_set = set()

    for i in range(len(nums) - 2):
        left, right = i + 1, len(nums) - 1

        while left < right:
            total = nums[i] + nums[left] + nums[right]
            if total == 0:
                result_set.add((nums[i], nums[left], nums[right]))
                left += 1
                right -= 1
            elif total < 0:
                left += 1
            else:
                right -= 1

    return [list(t) for t in result_set]


# Q6: Recursive approach (less efficient)
def threeSum_recursive(nums: list[int]) -> list[list[int]]:
    """
    Time Complexity: O(n^2)
    Space Complexity: O(n) for recursion

    Approach: Recursively reduce to 2Sum problem.
    """
    nums.sort()
    result = []

    def n_sum(nums, n, target, start=0, current=[]):
        if n == 2:
            left, right = start, len(nums) - 1
            while left < right:
                total = nums[left] + nums[right]
                if total == target:
                    result.append(current + [nums[left], nums[right]])
                    while left < right and nums[left] == nums[left + 1]:
                        left += 1
                    while left < right and nums[right] == nums[right - 1]:
                        right -= 1
                    left += 1
                    right -= 1
                elif total < target:
                    left += 1
                else:
                    right -= 1
            return

        for i in range(start, len(nums) - n + 1):
            if i > start and nums[i] == nums[i - 1]:
                continue
            n_sum(nums, n - 1, target - nums[i], i + 1, current + [nums[i]])

    n_sum(nums, 3, 0)
    return result


# Q7: Generator-based approach
def threeSum_generator(nums: list[int]) -> list[list[int]]:
    """
    Time Complexity: O(n^2)
    Space Complexity: O(n)

    Approach: Use generator for lazy evaluation.
    """
    nums.sort()
    result = []
    seen = set()

    def generate_triplets():
        for i in range(len(nums) - 2):
            if nums[i] in seen:
                continue
            seen.add(nums[i])

            left, right = i + 1, len(nums) - 1
            while left < right:
                total = nums[i] + nums[left] + nums[right]
                if total == 0:
                    yield (nums[i], nums[left], nums[right])
                    left += 1
                    right -= 1
                elif total < 0:
                    left += 1
                else:
                    right -= 1

    return [list(t) for t in generate_triplets()]


# Q8: Counting sort optimization
def threeSum_counting(nums: list[int]) -> list[list[int]]:
    """
    Time Complexity: O(n^2 + k) where k is range of numbers
    Space Complexity: O(k)

    Approach: Use counting for better performance on small range inputs.
    """
    nums.sort()
    result = []

    for i in range(len(nums) - 2):
        if i > 0 and nums[i] == nums[i - 1]:
            continue

        target = -nums[i]
        left, right = i + 1, len(nums) - 1

        while left < right:
            current_sum = nums[left] + nums[right]
            if current_sum == target:
                result.append([nums[i], nums[left], nums[right]])
                left += 1
                right -= 1
                while left < right and nums[left] == nums[left - 1]:
                    left += 1
                while left < right and nums[right] == nums[right + 1]:
                    right -= 1
            elif current_sum < target:
                left += 1
            else:
                right -= 1

    return result


# Q9: Using bisect for sorted operations
def threeSum_bisect(nums: list[int]) -> list[list[int]]:
    """
    Time Complexity: O(n^2)
    Space Complexity: O(1)

    Approach: Standard two-pointer after sorting.
    """
    import bisect

    nums.sort()
    result = []

    for i in range(len(nums) - 2):
        if i > 0 and nums[i] == nums[i - 1]:
            continue

        target = -nums[i]
        left, right = i + 1, len(nums) - 1

        while left < right:
            total = nums[left] + nums[right]
            if total == target:
                result.append([nums[i], nums[left], nums[right]])
                left += 1
                right -= 1
                while left < right and nums[left] == nums[left - 1]:
                    left += 1
                while left < right and nums[right] == nums[right + 1]:
                    right -= 1
            elif total < target:
                left += 1
            else:
                right -= 1

    return result


# Q10: Index-based approach
def threeSum_index(nums: list[int]) -> list[list[int]]:
    """
    Time Complexity: O(n^2)
    Space Complexity: O(1)

    Approach: Track indices instead of values for deduplication.
    """
    nums.sort()
    result = []
    n = len(nums)

    for i in range(n - 2):
        if i > 0 and nums[i] == nums[i - 1]:
            continue

        left = i + 1
        right = n - 1

        while left < right:
            total = nums[i] + nums[left] + nums[right]

            if total == 0:
                result.append([nums[i], nums[left], nums[right]])

                # Skip duplicates
                while left + 1 < right and nums[left] == nums[left + 1]:
                    left += 1
                while right - 1 > left and nums[right] == nums[right - 1]:
                    right -= 1

                left += 1
                right -= 1
            elif total < 0:
                left += 1
            else:
                right -= 1

    return result


if __name__ == "__main__":
    test_cases = [
        ([-1, 0, 1, 2, -1, -4], [[-1, -1, 2], [-1, 0, 1]]),
        ([0], []),
        ([-2, 0, 1, 1, 2], [[-2, 0, 2], [-2, 1, 1]]),
        ([0, 0, 0, 0], [[0, 0, 0]]),
        ([-1, -1, -1, 2], []),
        ([-4, -2, -2, -2, 0, 1, 2, 2, 2, 3, 3, 4, 4, 6, 6], [[-4, -2, 6], [-4, 0, 4], [-4, 1, 3], [-4, 2, 2], [-2, -2, 4], [-2, 0, 2], [0, 0, 0]]),
    ]

    functions = [
        threeSum_sorting,
        threeSum_hashset,
        threeSum_two_pointer,
        threeSum_dict,
        threeSum_set_dedup,
        threeSum_recursive,
        threeSum_generator,
        threeSum_counting,
        threeSum_bisect,
        threeSum_index,
    ]

    for i, func in enumerate(functions, 1):
        print(f"\nQ{i}: {func.__name__}")
        print("-" * 70)
        all_passed = True
        for nums, expected in test_cases:
            result = func(nums[:])
            result.sort()
            expected.sort()
            passed = result == expected
            all_passed = all_passed and passed
            status = "✓" if passed else "✗"
            print(f"  {status} threeSum({nums}) = {result}")
            if not passed:
                print(f"      Expected: {expected}")

        if all_passed:
            print(f"  All tests passed!")
